### ISSUE TITLE
Bookmark folder gets duplicated on desktop when edited on Android device while sync is in progress

### ISSUE THREAD

**Issue by srirambv** [brave/brave-browser#2569]
## Description 

Bookmark folder gets duplicated on desktop when edited on Android device while sync is in progress

## Devices
Device 1: Ubuntu running bc beta `0.59.8`
Device 2: Samsung tab running 1.0.71(Sync2) build
## Steps to Reproduce

   1. Create sync chain on desktop 
   2. Add an Android device to the sync chain
   3. Add a couple of bookmarks on desktop, ensure it sync to mobile
   4. Add a couple of bookmarks on mobile, ensure it syncs to desktop
   5. Import a bookmark html file on desktop ([brave_sorted_1000.zip](https://github.com/brave/brave-browser/files/2678972/brave_sorted_1000.zip))
   6. When empty root folder `Imported` is created, edit the HTML folder in Android and rename to `Imported HTML`
   7. Wait for sync to complete
   8. `Imported HTML` is the root folder on bookmarks bar and inside it has another `Imported` bookmark folder which was the root folder in step 5

## Actual result:

![screenshot from 2018-12-14 12-40-02](https://user-images.githubusercontent.com/17010094/49988629-7ef90f00-ff9d-11e8-96b6-278b998fd603.png)
Android has the same folder structre
## Expected result:
`Imported HTML` should contain one sub folder `Bookmarks Bar` which should further contain 2 nested bookmarks folder `0-99` and `Bookmarks Toolbar`.

## Reproduces how often: 

Easy

## Brave version (brave://version info)

Brave | 0.59.8 Chromium: 71.0.3578.98 (Official Build) beta(64-bit)
-- | --
Revision | 15234034d19b85dcd9a03b164ae89d04145d8368-refs/branch-heads/3578@{#897}
OS | Linux

### Reproducible on current release:
- Does it reproduce on brave-browser dev/beta builds? 
Yes
### Website problems only:  
- Does the issue resolve itself when disabling Brave Shields? NA
- Is the issue reproducible on the latest version of Chrome? NA

### Additional Information

cc: @darkdh @AlexeyBarabash @brave/legacy_qa

**Comment by AlexeyBarabash** [comment]
Hi, @srirambv .

I hadn't fully understand the STR.

`
6. When empty root folder Imported is created, edit the HTML folder in Android and rename to Imported HTML
`
I treated this step as `6. When empty root folder Imported is created on Android device, rename it "Imported" => "Imported HTML" in Android`


After step 5 I have this structure on desktop:
![screenshot from 2018-12-18 10-08-42__](https://user-images.githubusercontent.com/24739341/50142415-c7bd0a80-02b2-11e9-90d7-4176af4cb32d.png)

After step 8 I have this structure on desktop:
![screenshot from 2018-12-18 10-27-52__](https://user-images.githubusercontent.com/24739341/50142424-cd1a5500-02b2-11e9-9113-6dfdaee31596.png)

Root folder on desktop was renamed "Imported" => "Imported HTML" as a reflecting of changes on Android.
So I cannot understand what is the issue or it was fixed with a fix for some other issues.

Tested with Android `1.0.71(sync3)` and Desktop:
Brave	0.61.0 Chromium: 71.0.3578.98 (Developer Build) (64-bit)
Revision	15234034d19b85dcd9a03b164ae89d04145d8368-refs/branch-heads/3578@{#897}

**Comment by srirambv** [comment]
@AlexeyBarabash Here's what the imported bookmarks folder structure looks like when you import the bookmarks on bc 
```
[+] Root bookmark toolbar
   [-] Imported
     [-] Bookmarks Bar
        [-] 0-99
        [-] Bookmarks toolbar
           [-] 100-199
             [-] .....
```
Here's what it looks like after you edit the empty `Imported` bookmark folder on Android which is basically wrong.
```
[+] Root bookmark toolbar
   **[-] Imported HTML**
     **[-] Bookmarks Bar**
         [-] Imported
           [-] Bookmarks bar
              [-] 0-99
              [-] Bookmarks toolbar
                 [-] 100-199
                    [-] .....
```

On Android you don't create a new folder and move the bookmarks so it should just rename the existing `Imported` bookmark folder to `Imported HTML` folder. This is the folder structre am expecting after I edit the folder name on Android.
```
[+] Root bookmark toolbar
   [-] Imported HTML
     [-] Bookmarks Bar
        [-] 0-99
        [-] Bookmarks toolbar
           [-] 100-199
             [-] .....
```

Its creating a new root level folder for the `Imported` folder and pulling in the entire html file under it. 
Can see two possibility to overcome this
1. Do not allow editing bookmarks/folders on Android until the sync is complete (doubt this is possible)
2. Batch sync entire folder structure instead of just sending 50 items per sync call (current behaviour) so that even if the folder/bookmark is edited should avoid this issue

**Comment by srirambv** [comment]
@brave/legacy_qa needs to retest this once new beta build is out

**Comment by bhlevca** [comment]
I have similar problems with sync:

For example  on my bookmark bar I have 3 folders :+1: 
A   B  C

After sync form android I get
A A B B C C

The content of those folders is most of the time complete. I am not sure if I leave it working  the folders would be multiplied even further like  AAA BBB CCC  and so on. I disabled it for now until this bug is fixed.

I am sorry that I did not keep a screenshot, But I think that I can replicate it if needed.

**Comment by GeetaSarvadnya** [comment]
1+ https://community.brave.com/t/bookmarks-and-passwords-sync-issue/53566/4

One more scenario on duplicate bookmarks sync. 
1. Clean profile 0.61.52
2. Start sync chain on Desktop windows
3. Join sync chain through a QR code on Android
4. Import 1000 bookmarks on Desktop
5. Make sure 1000 bookmarks synced on Android
6. Reset sync chain on Desktop
7. Create a sync chain again between Desktop and Android ( copied sync code from Android)
8. Verify that duplicate bookmarks synced again on both Desktop and Android

**Comment by djsalmon** [comment]
running version Version 0.64.77 Chromium: 74.0.3729.169 (Official Build) (64-bit), the steps outlined by @GeetaSarvadnya are what I experienced.  This is a valid use case and outcome.

**Comment by kjozwiak** [comment]
@AlexeyBarabash we should probably take another look at this as we've received several +1 from the community.

**Comment by AlexeyBarabash** [comment]
@kjozwiak just re-checked the original STR and STR of two messages above from @GeetaSarvadnya , could not reproduce any of them.

**Comment by bhlevca** [comment]
@AlexeyBarabash I can reproduce it every time. You need to be a little patient and try to sync on both sides.  Perhaps a few times. Try to add on both sides. I can reproduce it even if it is not during sync.

**Comment by AlexeyBarabash** [comment]
@bhlevca are you talking about steps  from @srirambv or about steps  from @GeetaSarvadnya ?

**Comment by bhlevca** [comment]
@AlexeyBarabash I am closer to what @GeetaSarvadnya did.

**Comment by bhlevca** [comment]
@AlexeyBarabash  but I did not use the nightly, Just the standard download.

**Comment by leobel96** [comment]
Same @GeetaSarvadnya experience here

**Comment by bhlevca** [comment]
Thank you for closing the bug, but what's the need for the **wontfix** label if if was already fixed on May 28?

**Comment by jsecretan** [comment]
Sorry for the confusion @bhlevca, wanted to log for future filtering.  This will be fixed in the coming Brave Sync version 2 which you can find currently in Desktop nightly and Android nightly.
